streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.rocksdb.RocksIterator;
+
+import java.util.Comparator;
+import java.util.Set;
+
+class RocksDBPrefixIterator extends RocksDbIterator {
+    private byte[] rawPrefix;
+
+    RocksDBPrefixIterator(final String name,
+                          final RocksIterator newIterator,
+                          final Set<KeyValueIterator<Bytes, byte[]>> openIterators,
+                          final Bytes prefix) {
+        super(name, newIterator, openIterators);
+        rawPrefix = prefix.get();
+        newIterator.seek(rawPrefix);
+    }
+
+    @Override
+    public synchronized boolean hasNext() {
+        if (!super.hasNext()) {
+            return false;
+        }
+
+        final byte[] rawNextKey = super.peekNextKey().get();
+        for (int i = 0; i < rawPrefix.length; i++) {
+            if (i == rawNextKey.length) {
+                throw new ArrayIndexOutOfBoundsException("Unexpected RocksDB Key Value. Should have been skipped with seek.");
+            }
+            if (rawNextKey[i] != rawPrefix[i]) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
Huh, neat...
As I read it, the serial format for our CombinedKeys is [FK length][FK bytes][PK bytes]
This means that all the records with same-length foreign keys are sorted together, a prefix scan won't
suffer any ambiguity between FK and prefixes of the PK.
This seems to avoid a problem we face in the session window store. Putting it in terms of this key
If we have two records with r1.FK=aa r1.PK=b and r2.FK=a and r2.PK=ab,
they are both serialized as aab, and in particular,
a prefix scan would get pseudomatches that are out of the range and
have to handle it by decomposing the serial form and then double-checking the prefix.
But since you prefix by the size of the FK, there's no ambiguity, even if the FKs are variably sized!

But it is worth noting that this depends on the exact serialization format.

Reply:
Yep, this is the intended design. It definitely is sensitive to how the serialization is handled, but I think that it's okay as it currently stands.